src/main/java/software/amazon/smithy/lsp/project/ProjectManager.java

```diff
@@ -126,4 +155,51 @@ public Document getDocument(String uri) {
         }
         return project.getDocument(uri);
     }
+
+    /**
+     * Computes per-project file changes from the given file events.
+     *
+     * @param events The file events to compute per-project file changes from
+     * @return A map of project name to the corresponding project's changes
+     */
+    public Map<String, ProjectChanges> computeProjectChanges(List<FileEvent> events) {
+        // Note: we could eagerly compute these and store them, but project changes are relatively rare,
+        //  and doing it this way means we don't need to manage the state.
+        Map<String, PathMatcher> projectSmithyFileMatchers = new HashMap<>(nonDetachedProjects().size());
+        Map<String, PathMatcher> projectBuildFileMatchers = new HashMap<>(nonDetachedProjects().size());
+
+        Map<String, ProjectChanges> changes = new HashMap<>(nonDetachedProjects().size());
+
+        nonDetachedProjects().forEach((projectName, project) -> {
+            projectSmithyFileMatchers.put(projectName, ProjectFilePatterns.getSmithyFilesPathMatcher(project));
+            projectBuildFileMatchers.put(projectName, ProjectFilePatterns.getBuildFilesPathMatcher(project));
+
+            // Need these to be hash sets so they are mutable
+            changes.put(projectName, new ProjectChanges(new HashSet<>(), new HashSet<>(), new HashSet<>()));
+        });
+
+        for (FileEvent event : events) {
+            String changedUri = event.getUri();
+            Path changedPath = Path.of(LspAdapter.toPath(changedUri));
+            if (changedUri.endsWith(".smithy")) {
+                projectSmithyFileMatchers.forEach((projectName, matcher) -> {
+                    if (matcher.matches(changedPath)) {
+                        if (event.getType() == FileChangeType.Created) {
+                            changes.get(projectName).createdSmithyFileUris().add(changedUri);
```

Comment:
Thought: What if we did "getOrCompute" here? Then we could only create the changes for projects that have them.

Reply:
No, the hashsets are already initialized

Reply:
Note: if you have lots of projects, this will create a bunch of garbage because most times you aren't getting multiple sets of large updates to a project

Reply:
Added a note